src/samplers/v2/sampling_state.js

```diff
@@ -0,0 +1,143 @@
+export const WARN_CANNOT_REVERT_FINAL = 'The final state cannot be reverted.';
+// export const WARN_CANNOT_UNSAMPLE = 'The final state cannot be reverted.';
+// export const WARN_CANNOT_REMOVE_DEBUG = 'The debug flag cannot be reverted.';
+// export const WARN_CANNOT_REMOVE_FIREHOSE = 'The firehose flag cannot be reverted.';
+
+export default class SamplingState {
+  _extendedState: { [string]: any } = {};
+  _flags: number = 0;
+  _isFinal: boolean = false;
+  _isFinalizable: boolean = true;
+  _localRootSpanIdStr: ?string;
+
+  constructor(localRootSpanIdStr: ?string) {
+    this._localRootSpanIdStr = localRootSpanIdStr;
+  }
+
+  isLocalRootSpan(context: SpanContext) {
+    return this._localRootSpanIdStr === context.spanIdStr;
+  }
+
+  extendedState() {
+    return this._extendedState;
+  }
+
+  localRootSpanId() {
+    return this._localRootSpanIdStr;
+  }
+
+  isFinalizable() {
+    return this._isFinalizable;
+  }
+
+  setIsFinalizable(value: false): ?string {
+    if (value) {
+      if (!this._isFinalizable) {
+        return WARN_CANNOT_REVERT_FINALIZABLE;
+      }
+      return WARN_SHOULD_NOT_ENABLE_FINALIZABLE;
+    }
+    this._isFinalizable = false;
+    this._isFinal = false;
+  }
+
+  isFinal() {
+    return this._isFinal;
+  }
+
+  setIsFinal(value: boolean): ?string {
+    if (!this._isFinalizable) {
+      return WARN_NOT_FINALIZABLE;
+    }
+    // TODO(joe): verify this is the behavior we want
+    if (!value && this._isFinal) {
+      return WARN_CANNOT_REVERT_FINAL;
+    }
+    this._isFinal = value;
+  }
+
+  isSampled() {
+    return Boolean(this._flags & SAMPLED_MASK);
+  }
+
+  // setIsSampled(value: true) {
+  //   if (!value && this._isSampled) {
+  //     return WARN_CANNOT_UNSAMPLE;
+  //   }
+  //   this._toggleFlag(SAMPLED_MASK, enable);
+  // }
+  setIsSampled(enable: boolean) {
+    this._toggleFlag(SAMPLED_MASK, enable);
+  }
+
+  isDebug() {
+    return Boolean(this._flags & DEBUG_MASK);
+  }
+
+  // setIsDebug(value: true) {
+  //   if (!value && this.isDebug()) {
+  //     return WARN_CANNOT_REMOVE_DEBUG;
+  //   }
+  //   this._toggleFlag(DEBUG_MASK, value);
+  // }
+  setIsDebug(enable: boolean) {
+    this._toggleFlag(DEBUG_MASK, enable);
+  }
+
+  isFirehose() {
+    return Boolean(this._flags & FIREHOSE_MASK);
+  }
+
+  // setIsFirehose(value: true) {
+  //   if (!value && this._isFirehose) {
+  //     return WARN_CANNOT_REMOVE_FIREHOSE;
+  //   }
+  //   this._toggleFlag(FIREHOSE_MASK, enable);
+  // }
+  setIsFirehose(enable: boolean) {
+    this._toggleFlag(FIREHOSE_MASK, enable);
+  }
+
+  flags() {
+    return this._flags;
+  }
+
+  setFlags(flags: number) {
+    this._flags = flags;
+  }
+
+  _toggleFlag(mask: number, enable: boolean) {
+    if (enable) {
+      this._flags |= mask;
+    } else {
+      this._flags &= ~mask;
+    }
+  }
+
+  reset() {
+    this._extendedState = {};
+    this._flags = 0;
```

Comment:
@yurishkuro @vprithvi

Seems like this needs to just reset whether or not it's sampled (but not the rest of the flags) only if doesn't have the debug flag?

Reply:
I'm not sure I follow - what is the scenario you are thinking of?

Reply:
For instance, deferred sampling when setting the operation name.